src/server/scripts/Outland/CoilfangReservoir/SerpentShrine/boss_leotheras_the_blind.cpp

```diff
@@ -80,267 +75,235 @@ const Position channelersPos[MAX_CHANNELERS] =
 
-            if (me->GetDisplayId() == me->GetNativeDisplayId())
-                DoMeleeAttackIfReady();
-            else if (me->isAttackReady(BASE_ATTACK))
-            {
-                me->CastSpell(me->GetVictim(), SPELL_CHAOS_BLAST, false);
-                me->setAttackTimer(BASE_ATTACK, 2000);
-            }
+        if (me->GetDisplayId() == me->GetNativeDisplayId())
+        {
+            DoMeleeAttackIfReady();
         }
-    };
+        else if (me->isAttackReady(BASE_ATTACK))
+        {
+            me->CastSpell(me->GetVictim(), SPELL_CHAOS_BLAST, false);
+            me->setAttackTimer(BASE_ATTACK, 2000);
+        }
+    }
+private:
+    bool _recentlySpoken;
 };
 
-class npc_inner_demon : public CreatureScript
+struct npc_inner_demon : public ScriptedAI
 {
-public:
-    npc_inner_demon() : CreatureScript("npc_inner_demon") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
+    npc_inner_demon(Creature* creature) : ScriptedAI(creature)
     {
-        return GetSerpentShrineAI<npc_inner_demonAI>(creature);
+        _scheduler.SetValidator([this]
+        {
+            return !me->HasUnitState(UNIT_STATE_CASTING);
+        });
+
+        instance = creature->GetInstanceScript();
     }
 
-    struct npc_inner_demonAI : public ScriptedAI
-    {
-        npc_inner_demonAI(Creature* creature) : ScriptedAI(creature)
-        {
-        }
+    InstanceScript* instance;
 
-        ObjectGuid ownerGUID;
-        EventMap events;
+    void EnterEvadeMode(EvadeReason /*why*/) override
+    {
+        me->DespawnOrUnsummon(1);
+    }
 
-        void EnterEvadeMode(EvadeReason /*why*/) override
-        {
-            me->DespawnOrUnsummon(1);
-        }
+    void IsSummonedBy(WorldObject* summoner) override
+    {
+        if (!summoner)
+            return;
 
-        void IsSummonedBy(WorldObject* summoner) override
+        _scheduler.CancelAll();
+        _scheduler.Schedule(4s, [this](TaskContext context)
         {
-            if (!summoner)
-                return;
-
-            ownerGUID = summoner->GetGUID();
-            events.Reset();
-            events.ScheduleEvent(EVENT_SPELL_SHADOW_BOLT, 4000);
-        }
+            DoCastVictim(SPELL_SHADOW_BOLT);
+            context.Repeat(6s);
+        });
+    }
 
-        void JustDied(Unit* /*killer*/) override
+    void JustDied(Unit* /*killer*/) override
+    {
+        if (Creature* leotheras = instance->GetCreature(DATA_LEOTHERAS_THE_BLIND))
         {
-            if (Unit* unit = ObjectAccessor::GetUnit(*me, ownerGUID))
-                unit->RemoveAurasDueToSpell(SPELL_INSIDIOUS_WHISPER);
+            leotheras->RemoveAurasDueToSpell(SPELL_INSIDIOUS_WHISPER);
         }
+    }
 
-        void DamageTaken(Unit* who, uint32& damage, DamageEffectType, SpellSchoolMask) override
+    void DamageTaken(Unit* who, uint32& damage, DamageEffectType, SpellSchoolMask) override
+    {
+        if(Creature* leotheras = instance->GetCreature(DATA_LEOTHERAS_THE_BLIND))
```

Comment:
```suggestion
        if (Creature* leotheras = instance->GetCreature(DATA_LEOTHERAS_THE_BLIND))
```

A few others below as well